File: src/python/src/civic_exchange_protocol/entity/identifiers.py
"""Entity identifier types for CEP.

CEP supports multiple identifier schemes organized into tiers:

- Tier 1 (Global): LEI (Legal Entity Identifier)
- Tier 2 (Federal): SAM.gov UEI
- Tier 3 (Sub-National): SNFEI (generated hash-based identifier)
- Extended: Canadian BN, UK Companies House, etc.
"""

from dataclasses import dataclass
import json
from typing import Optional

from civic_exchange_protocol.core import Canonicalize, insert_if_present
from civic_exchange_protocol.snfei import Snfei


@dataclass(frozen=True)
class SamUei:
    """SAM.gov Unique Entity Identifier (12 alphanumeric characters)."""

    value: str

    def __post_init__(self) -> None:
        """Validate the SAM UEI format after initialization."""
        if not self._is_valid(self.value):
            raise ValueError(f"Invalid SAM UEI: {self.value}")

    @staticmethod
    def _is_valid(value: str) -> bool:
        return (
            len(value) == 12 and all(c.isupper() or c.isdigit() for c in value) and value.isalnum()
        )

    @classmethod
    def new(cls, value: str) -> Optional["SamUei"]:
        """Create a new SAM UEI, returning None if invalid."""
        try:
            return cls(value)
        except ValueError:
            return None

    def as_str(self) -> str:
        """Return the SAM UEI as a string."""
        return self.value


@dataclass(frozen=True)
class Lei:
    """Legal Entity Identifier per ISO 17442 (20 alphanumeric characters)."""

    value: str

    def __post_init__(self) -> None:
        """Validate the LEI format after initialization."""
        if not self._is_valid(self.value):
            raise ValueError(f"Invalid LEI: {self.value}")

    @staticmethod
    def _is_valid(value: str) -> bool:
        return len(value) == 20 and value.isalnum()

    @classmethod
    def new(cls, value: str) -> Optional["Lei"]:
        """Create a new LEI, returning None if invalid."""
        try:
            return cls(value.upper())
        except ValueError:
            return None

    def as_str(self) -> str:
        """Return the LEI as a string."""
        return self.value


@dataclass(frozen=True)
class CanadianBn:
    """Canadian Business Number with program account."""

    value: str

    def __post_init__(self) -> None:
        """Validate the Canadian BN format after initialization."""
        if not self._is_valid(self.value):
            raise ValueError(f"Invalid Canadian BN: {self.value}")

    @staticmethod
    def _is_valid(value: str) -> bool:
        # Pattern: 9 digits + 2 letters + 4 digits (e.g., 123456789RC0001)
        if len(value) != 15:
            return False
        digits1 = value[:9]
        letters = value[9:11]
        digits2 = value[11:15]
        return digits1.isdigit() and letters.isalpha() and letters.isupper() and digits2.isdigit()

    @classmethod
    def new(cls, value: str) -> Optional["CanadianBn"]:
        """Create a new Canadian BN, returning None if invalid."""
        try:
            return cls(value)
        except ValueError:
            return None

    def as_str(self) -> str:
        """Return the Canadian BN as a string."""
        return self.value


@dataclass
class AdditionalScheme:
    """An additional identifier scheme not explicitly defined in the schema."""

    scheme_uri: str
    value: str


@dataclass
class EntityIdentifiers(Canonicalize):
    """Collection of all known identifiers for an entity."""

    sam_uei: SamUei | None = None
    lei: Lei | None = None
    snfei: Snfei | None = None
    canadian_bn: CanadianBn | None = None
    additional_schemes: list[AdditionalScheme] | None = None

    def with_sam_uei(self, uei: SamUei) -> "EntityIdentifiers":
        """Return a new EntityIdentifiers with the SAM UEI set."""
        return EntityIdentifiers(
            sam_uei=uei,
            lei=self.lei,
            snfei=self.snfei,
            canadian_bn=self.canadian_bn,
            additional_schemes=self.additional_schemes,
        )

    def with_lei(self, lei: Lei) -> "EntityIdentifiers":
        """Return a new EntityIdentifiers with the LEI set."""
        return EntityIdentifiers(
            sam_uei=self.sam_uei,
            lei=lei,
            snfei=self.snfei,
            canadian_bn=self.canadian_bn,
            additional_schemes=self.additional_schemes,
        )

    def with_snfei(self, snfei: Snfei) -> "EntityIdentifiers":
        """Return a new EntityIdentifiers with the SNFEI set."""
        return EntityIdentifiers(
            sam_uei=self.sam_uei,
            lei=self.lei,
            snfei=snfei,
            canadian_bn=self.canadian_bn,
            additional_schemes=self.additional_schemes,
        )

    def has_any(self) -> bool:
        """Return True if at least one identifier is present."""
        return (
            self.sam_uei is not None
            or self.lei is not None
            or self.snfei is not None
            or self.canadian_bn is not None
            or (self.additional_schemes is not None and len(self.additional_schemes) > 0)
        )

    def primary_identifier(self) -> str | None:
        """Return the 'best' identifier for use as the verifiable ID.

        Priority: LEI > SAM UEI > SNFEI > Canadian BN > first additional
        """
        if self.lei is not None:
            return f"cep-entity:lei:{self.lei.as_str()}"
        if self.sam_uei is not None:
            return f"cep-entity:sam-uei:{self.sam_uei.as_str()}"
        if self.snfei is not None:
            return f"cep-entity:snfei:{self.snfei.as_str()}"
        if self.canadian_bn is not None:
            return f"cep-entity:canadian-bn:{self.canadian_bn.as_str()}"
        if self.additional_schemes and len(self.additional_schemes) > 0:
            return f"cep-entity:other:{self.additional_schemes[0].value}"
        return None

    def canonical_fields(self) -> dict[str, str]:
        """Return the canonical fields in alphabetical order."""
        fields: dict[str, str] = {}

        # Additional schemes serialized as JSON array string
        if self.additional_schemes and len(self.additional_schemes) > 0:
            sorted_schemes = sorted(self.additional_schemes, key=lambda x: x.scheme_uri)
            schemes_data = [{"schemeUri": s.scheme_uri, "value": s.value} for s in sorted_schemes]
            fields["additionalSchemes"] = json.dumps(schemes_data, separators=(",", ":"))

        insert_if_present(
            fields, "canadianBn", self.canadian_bn.as_str() if self.canadian_bn else None
        )
        insert_if_present(fields, "lei", self.lei.as_str() if self.lei else None)
        insert_if_present(fields, "samUei", self.sam_uei.as_str() if self.sam_uei else None)
        insert_if_present(fields, "snfei", self.snfei.as_str() if self.snfei else None)

        return fields

File: src/rust/cep-entity/src/identifiers.rs
//! Entity identifier types for CEP.

use cep_core::canonical::{insert_if_present, Canonicalize};
use serde::{Deserialize, Serialize};
use sha2::{Digest, Sha256};
use std::collections::BTreeMap;

/// SAM.gov Unique Entity Identifier (12 alphanumeric characters).
#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub struct SamUei(String);

impl SamUei {
    pub fn new(value: &str) -> Option<Self> {
        if value.len() == 12 && value.chars().all(|c| c.is_ascii_uppercase() || c.is_ascii_digit()) {
            Some(Self(value.to_string()))
        } else {
            None
        }
    }

    pub fn as_str(&self) -> &str {
        &self.0
    }
}

/// Legal Entity Identifier per ISO 17442 (20 alphanumeric characters).
#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub struct Lei(String);

impl Lei {
    pub fn new(value: &str) -> Option<Self> {
        if value.len() == 20 && value.chars().all(|c| c.is_ascii_alphanumeric()) {
            Some(Self(value.to_uppercase()))
        } else {
            None
        }
    }

    pub fn as_str(&self) -> &str {
        &self.0
    }
}

/// Sub-National Federated Entity Identifier (SHA-256 hash, 64 hex chars).
#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub struct Snfei(String);

impl Snfei {
    pub fn from_hash(hash: &str) -> Option<Self> {
        if hash.len() == 64 && hash.chars().all(|c| c.is_ascii_hexdigit()) {
            Some(Self(hash.to_lowercase()))
        } else {
            None
        }
    }

    pub fn generate(normalized_name: &str, jurisdiction_iso: &str) -> Self {
        let input = format!("{}|{}", normalized_name, jurisdiction_iso);
        let mut hasher = Sha256::new();
        hasher.update(input.as_bytes());
        let result = hasher.finalize();
        Self(format!("{:x}", result))
    }

    pub fn as_str(&self) -> &str {
        &self.0
    }
}

/// Canadian Business Number with program account.
#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub struct CanadianBn(String);

impl CanadianBn {
    pub fn new(value: &str) -> Option<Self> {
        if value.len() == 15 {
            let (digits1, rest) = value.split_at(9);
            let (letters, digits2) = rest.split_at(2);
            if digits1.chars().all(|c| c.is_ascii_digit())
                && letters.chars().all(|c| c.is_ascii_uppercase())
                && digits2.chars().all(|c| c.is_ascii_digit())
            {
                return Some(Self(value.to_string()));
            }
        }
        None
    }

    pub fn as_str(&self) -> &str {
        &self.0
    }
}

/// An additional identifier scheme.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AdditionalScheme {
    pub scheme_uri: String,
    pub value: String,
}

/// Collection of all known identifiers for an entity.
#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct EntityIdentifiers {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub sam_uei: Option<SamUei>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub lei: Option<Lei>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub snfei: Option<Snfei>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub canadian_bn: Option<CanadianBn>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub additional_schemes: Option<Vec<AdditionalScheme>>,
}

impl EntityIdentifiers {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn with_sam_uei(mut self, uei: SamUei) -> Self {
        self.sam_uei = Some(uei);
        self
    }

    pub fn with_lei(mut self, lei: Lei) -> Self {
        self.lei = Some(lei);
        self
    }

    pub fn with_snfei(mut self, snfei: Snfei) -> Self {
        self.snfei = Some(snfei);
        self
    }

    pub fn has_any(&self) -> bool {
        self.sam_uei.is_some()
            || self.lei.is_some()
            || self.snfei.is_some()
            || self.canadian_bn.is_some()
            || self.additional_schemes.as_ref().map_or(false, |v| !v.is_empty())
    }

    pub fn primary_identifier(&self) -> Option<String> {
        if let Some(ref lei) = self.lei {
            return Some(format!("cep-entity:lei:{}", lei.as_str()));
        }
        if let Some(ref uei) = self.sam_uei {
            return Some(format!("cep-entity:sam-uei:{}", uei.as_str()));
        }
        if let Some(ref snfei) = self.snfei {
            return Some(format!("cep-entity:snfei:{}", snfei.as_str()));
        }
        if let Some(ref bn) = self.canadian_bn {
            return Some(format!("cep-entity:canadian-bn:{}", bn.as_str()));
        }
        None
    }
}

impl Canonicalize for EntityIdentifiers {
    fn canonical_fields(&self) -> BTreeMap<String, String> {
        let mut map = BTreeMap::new();
        if let Some(ref schemes) = self.additional_schemes {
            if !schemes.is_empty() {
                let mut sorted: Vec<_> = schemes.iter().collect();
                sorted.sort_by(|a, b| a.scheme_uri.cmp(&b.scheme_uri));
                let json = serde_json::to_string(&sorted).unwrap_or_default();
                map.insert("additionalSchemes".to_string(), json);
            }
        }
        insert_if_present(&mut map, "canadianBn", self.canadian_bn.as_ref().map(|x| x.as_str()));
        insert_if_present(&mut map, "lei", self.lei.as_ref().map(|x| x.as_str()));
        insert_if_present(&mut map, "samUei", self.sam_uei.as_ref().map(|x| x.as_str()));
        insert_if_present(&mut map, "snfei", self.snfei.as_ref().map(|x| x.as_str()));
        map
    }
}